### PR TITLE
Makes Ubuntu Xenial Xerus use PHP 5.6 instead of 5.5.

### PR DESCRIPTION
**Pull request by fretb**
As I was converging a testnode with Xenial, I noticed php 5.5 has some dependency hell when trying to install php5.5 (concerning php-apcu and php-yac). As icingaweb2 is compatible with PHP 5.6, I see no issue updating this.

### CODE REVIEW

**Comment by vkhatri** [comment]
@fretb Will work on it later today.

**Comment by fretb** [comment]
@vkhatri any updates?

**Comment by vkhatri** [comment]
@fretb My bad, I just merged #246. But, thank you for the PR though.

**Comment by Thomas-Gelf** [comment]
Hi @vkhatri,

may I ask why we do not use PHP 7.0? It is available on Xenial, much faster, uses about half the memory, and even more important: it's supported also upstream. PHP 5.5 is completely dead, 5.6 is no longer under active support and gets security fixes only.

**TL;DR** we work hard to support all those versions, you can opt for 5.6 with our full support. But please consider using 7.x. It works fine, is tested, has only advantages. **/TL;DR**

In Icinga Web 2 we support "all" versions from 5.3 up to 7.1, so there is nothing wrong with using 5.x - we support it. We recently changed some of our projects to run all the tests on all those versions, other projects will follow. See [Businessprocess](https://travis-ci.org/Icinga/icingaweb2-module-businessprocess/builds/191725912) or [Director](https://travis-ci.org/Icinga/icingaweb2-module-director/builds/192411829) for an example of how this looks like on Travis.

In addition to that, we also test our mostly used distributions. Those are not available on Travis, so please have a look [here](https://github.com/Icinga/icingaweb2-module-director/blob/master/doc/93-Testing.md#tests-for-supported-platforms) to get an idea of what I'm talking about.

We added CentOS 6 since then, as they (and RHEL 6) are running a very special variant of 5.3.x. Sooner or later we'll fade out those 5.x versions, this will allow us to write more efficient code. But don't worry, this will happen in a smooth way.

Cheers,
Thomas

**Comment by Datamancer1110** [comment]
I know I'm kind of a random guy here, but I _was_ the one that got PHP working in Xenial for Icinga (just updated dependencies, nothing terribly impressive). I will say that it concerns me somewhat that we have Trusty using PHP 5.5 and Xenial using PHP 5.6 (dunno about RHEL). If we wanted to, it looks like we could update this thing so that Trusty and Xenial (and probably RHEL too) use PHP7. 
If somebody important gives the green light, I'm happy to make the modifications and do some light local testing...this sort of change usually begets a full regression, however, and I do not have the time/resources to do that.

Let me know.

**Comment by Thomas-Gelf** [comment]
Hi @Datamancer1110,

thanks for your feedback and for your efforts! Just curious, what has been the problem at the time being? Is it in the packages shipped for Icinga Web 2? Because, honestly - I never used them, I usually run Icinga Web 2 directly from a GIT working tree :tongue: But if things are going wrong there I'd really love to be made aware of those issues.

In case we require a specific PHP version in the packages that should IMO be considered a bug. In case we don't, every available PHP version should be fine. And running PHP 7 on Xenial requires (as far as I can recall it right now):

```
apt-get install libapache2-mod-php7.0 php7.0-mysql [..]
a2dismod mpm_worker
a2dismod mpm_event
a2enmod mpm_prefork
systemctl restart apache2.service 
```

Alternatively, it's perfectly legal to run it via FPM - but as far as I remember Debian people decided to stay with mod-php. The problem is that usually the "wrong" MPM is enabled per default, and packagers do not switch it in their post-install scripts (what might be a wise decision. However, this then often leads to people ending up with "a white page" and no idea how to debug the problem.

Please give me some more context, let me know what you have been struggling with. I'm very interested in finding a good solution for everybody in this context.

Thanks,
Thomas

**Comment by Datamancer1110** [comment]
I'm not struggling with anything - however, I get the feeling that I was on the bleeding edge when I originally made Icinga2 (via Chef) work in Xenial. The problem I had run into was that, originally, we had one package list that needed to be installed (via Aptitude) when a system was detected as being Debian (via the recipe "server_os_packages.rb"). Conceptually, there's no problem there, but the issue arose that Xenial didn't know how to install the packaged called "php5" (among others), so I put in a case-switch block in based on Debian flavors, and renamed the packages to php5.5 (etc). Then, it seems like Xenial dropped PHP5.5 support altogether, so I updated the packages to be called php5.6...but I only made the change for Xenial - Trusty is still installing "php5" (which I believe is still php 5.5, though I'm not sure).

The concern I aired is that Icinga2 will use different versions of PHP depending on your OS. This...might end poorly for us, if some incompatibility gets coded in or something. If you are implying that PHP 7 works across the board, we should update server_os_packages to install PHP7 on All Deb/Fed systems, instead of having split versions. If you would like, I can make that modification, but I'd like clearance from an official member of the Icinga Team (I'm not sure, are you an official member?) before I make such a major change - especially since I don't have the resources to test it on Trusty - I would only check that it works well on Xenial.

As for the cookbooks themselves, they all work fine on Xenial since I made the changes I made this week. Just checking to see if we are worried about version parity, and/or if we want to upgrade PHP. If we do, I can code that for you guys.

**Comment by Thomas-Gelf** [comment]
> The concern I aired is that Icinga2 will use different versions of PHP depending on your OS. This...might end poorly for us, if some incompatibility gets coded in or something

Won't happen unless we discontinue older PHP versions. Such a move would be announced in time and won't happen without tagging a new (major?) version. We might slowly start to raise requirements for some modules over time to softly steer people towards newer PHP versions. Versions before 5.6 are [end of life](http://php.net/supported-versions.php) and do not even get security upgrades.

However, we still support PHP 5.3, even if it doesn't make fun to code for such outdated platforms. But hey, PHP 5.3 is still default on RHEL 6. Sad, but that's how "Enterprise" works. As you can see in the test runs linked above we test for all those versions, and we fully support them.

Back to Xenial: I would always use the default of the given distribution, and for Xenial this clearly is PHP 7. That's what you get when installing [php](http://packages.ubuntu.com/xenial/php). I see that you're using PHP 5.6 from some PPA. As a system operator I would not expect a recipe for Icinga to throw in PPAs with outdated PHP versions. They might conflict with what's already running with my system.

Similar thoughts related to the chosen packages: php5.6-fpm and libapache2-mod-php5.6 - they are contradicting, as usually one uses one or the other. As far as it goes for the Icinga project, we support quite some different ways of running Icinga Web 2: with standalone PHP using the embedded web server through the icingacli, with Apache and NGINX, using mod-php and FPM.

So, you have the free choice. In your position I'd opt for going the path prepared by packagers. As far as I know Debian packages default to libapache2. Guess they are not happy with that choice, as from what I heard they leave Xenial in an unusable state after installing the package - it's up to the user to get hist MPMs right. Not sure whether this information is still up to date.

Now, we as a project could also switch to propose FPM as a default. This would also allow to eventually use multiple PHP versions on the same system. Please wait before changing everything once again, I'll try to get more feedback from other team members.

Cheers,
Thomas

**Comment by Datamancer1110** [comment]
Sounds good, and I don't disagree...my process wasn't particularly...strategic.
1) We fired up an Ubuntu 16.04 box, as my company decided to move from 14.04 onto 16.04
2) I ran Chef to install our server software.
3) (among other things) Icigna blew up - Aptitude doesn't know what "php5" is
4) I popped onto one of our Ubuntu 14 boxes which already had Icinga on and checked the PHP version - it was 5.5
5) I figured out how to put PHP 5.5 onto Ubuntu 16. Maximum compatibility was my only goal.
6) Later, when PHP 5.5 support got dropped, I went to the next-closest version, again, to hopefully maintain compatibility.
7) As for _which_ packages get installed - that wasn't decided by me. Those were the packages Icinga already installed (via Chef) - I merely bumped the version numbers
8) I also updated, by necessity, where Icinga would look for the PHP INI file to configure date.timezone

So I'm not over here making philosophical choices (in fact, this conversation about PHP7 is about as philosophical as I've ever gotten about Icinga), I just needed something that would work. If you would have it differently, I'm all for it, and I'll even help code it if you'd like. I have no dog in the fight other than to say that I need Icinga2 functional and stable, as we use it in Production. (Also probably worth mentioning, I'm not a PHP dev, and frankly know very little about it, other than I don't like its syntax very much. I've worked with PHP just enough to get by, and no more, so I _really_ don't have an opinion on versions or w/e else)

**Comment by Thomas-Gelf** [comment]
Forget about philosophy, go with what the distribution gives you - this is PHP7 for Xenial ;-)

Cheers,
Thomas

**Comment by vkhatri** [comment]
Closing PR and tracking under issue #269
